File: test/leads.test.js
var assert = require('assert'),
    _ = require('lodash'),
    marketo = require('./helper/connection');

describe('Leads', function() {
  describe('#byId', function() {
    it('finds a lead by id only', function(done) {
      marketo.lead.byId(1).spread(function(response) {
        assert.equal(response.result.length, 1);
        assert.equal(response.result[0].id, 1);
        assert(_.has(response.result[0], 'email'));
        assert(_.has(response.result[0], 'lastName'));
        done();
      });
    });

    it('finds a lead but only retrieve the email field', function(done) {
      marketo.lead.byId(1, {fields: ['email']}).spread(function(resp) {
        assert.equal(resp.result.length, 1);

        var lead = resp.result[0];
        assert.equal(lead.id, 1);
        assert.equal(_.keys(lead).length, 2);
        assert(_.has(lead, 'email'));
        assert(!_.has(lead, 'lastName'));
        done();
      });
    });

    it('finds a lead and retrieve a subset of fields using csv', function(done) {
      marketo.lead.byId(1, {fields: 'email,lastName'}).spread(function(resp) {
        assert.equal(resp.result.length, 1);

        var lead = resp.result[0];
        assert.equal(lead.id, 1);
        assert.equal(_.keys(lead).length, 3);
        assert(_.has(lead, 'email'));
        assert(_.has(lead, 'lastName'));
        done();
      });
    });
  });

  describe('#find', function() {
    it('uses a single filter value', function(done) {
      marketo.lead.find('id', [1]).spread(function(resp) {
        assert.equal(resp.result.length, 1);

        var lead = resp.result[0];
        assert.equal(lead.id, 1);
        assert(_.has(lead, 'email'));
        assert(_.has(lead, 'lastName'));
        done();
      });
    });

    it('uses multiple filter values', function(done) {
      marketo.lead.find('id', [1,2]).spread(function(resp) {
        assert.equal(resp.result.length, 2);
        assert.equal(resp.result[0].id, 1);
        assert.equal(resp.result[1].id, 2);
        done();
      });
    });


    it('uses multiple filter values and retrieve a subset of fields', function(done) {
      marketo.lead.find('id', [1,2], {fields: ['email', 'lastName']}).spread(function(resp) {
        assert.equal(resp.result.length, 2);
        assert.equal(resp.result[0].id, 1);
        assert.equal(resp.result[1].id, 2);

        var lead = resp.result[0];
        assert.equal(lead.id, 1);
        assert(_.has(lead, 'email'));
        assert(_.has(lead, 'lastName'));
        done();
      });
    });
  });

  describe('#createOrUpdate', function() {
    it('updates an existing record by using an email', function(done) {
      marketo.lead.createOrUpdate([{email: 'ak+marketo1@usermind.com'}], {lookupField: 'email'})
        .spread(function(resp) {
          assert.equal(resp.result.length, 1);
          assert.equal(resp.result[0].id, 1);
          assert.equal(resp.result[0].status, 'updated');
          done();
        });
    })
  });
});

File: src.net/Marketo-Rest.Tests/LeadsTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Threading.Tasks;

namespace Marketo.Tests
{
    [TestClass, TestCategory("Leads")]
    public class LeadsTest
    {
        readonly MarketoClient marketo = Helper.config.Connection();

        [TestMethod]
        public async Task Leads_ById()
        {
            // finds a lead by id only
            var res = await marketo.Lead.ById(1);
            Assert.AreEqual(1, (int)res.result.length);

            var lead = res.result[0];
            Assert.AreEqual(1, (int)lead.id);
            Assert.IsNotNull((string)lead.email);
            Assert.IsNotNull((string)lead.lastName);

            // finds a lead but only retrieve the email field
            res = await marketo.Lead.ById(1, new { fields = new[] { "email" } });
            Assert.AreEqual(1, (int)res.result.length);

            lead = res.result[0];
            Assert.AreEqual(1, (int)lead.id);
            //Assert.AreEqual(2, lead.keys().length);
            Assert.IsNotNull((string)lead.email);
            Assert.IsNotNull((string)lead.lastName);


            // finds a lead and retrieve a subset of fields using csv
            res = await marketo.Lead.ById(1, new { fields = "email,lastName" });
            Assert.AreEqual(1, (int)res.result.length);

            lead = res.result[0];
            Assert.AreEqual(1, (int)lead.id);
            //Assert.AreEqual(3, lead.keys().length);
            Assert.IsNotNull((string)lead.email);
            Assert.IsNotNull((string)lead.lastName);
        }


        [TestMethod]
        public async Task Leads_Find()
        {
            // uses a single filter value
            var res = await marketo.Lead.Find("id", new object[] { 1 });
            Assert.AreEqual(1, (int)res.result.length);

            var lead = res.result[0];
            Assert.AreEqual(1, (int)lead.id);
            Assert.IsNotNull((string)lead.email);
            Assert.IsNotNull((string)lead.lastName);

            // uses multiple filter values
            res = await marketo.Lead.Find("id", new object[] { 1, 2 });
            Assert.AreEqual(2, (int)res.result.length);
            Assert.AreEqual(1, (int)res.result[0].id);
            Assert.AreEqual(2, (int)res.result[1].id);

            // uses multiple filter values and retrieve a subset of fields
            res = await marketo.Lead.Find("id", new object[] { 1, 2 }, new { fields = new[] { "email", "lastName" } });
            Assert.AreEqual(2, (int)res.result.length);
            Assert.AreEqual(1, (int)res.result[0].id);
            Assert.AreEqual(2, (int)res.result[1].id);

            lead = res.result[0];
            Assert.AreEqual(1, (int)lead.id);
            Assert.IsNotNull((string)lead.email);
            Assert.IsNotNull((string)lead.lastName);
        }
    }
}
